pkg/database/database.go

```diff
@@ -1145,18 +1171,188 @@ func (d *db) serializeTx(tx *store.Tx, spec *schema.EntriesSpec, snap *store.Sna
 	return stx, nil
 }
 
-func (d *db) ExportTxByID(req *schema.ExportTxRequest) ([]byte, error) {
+func (d *db) mayUpdateFollowerState(committedTxID uint64, newFollowerState *schema.FollowerState) error {
+	// clean up followerStates
+	// it's safe to remove up to latest tx committed in master
+	for uuid, st := range d.followerStates {
+		if st.precommittedTxID <= committedTxID {
+			delete(d.followerStates, uuid)
+		}
+	}
+
+	if newFollowerState.PrecommittedTxID <= committedTxID {
+		// as far as the master is concerned, nothing really new has happened
+		return nil
+	}
+
+	newFollowerAlh := schema.DigestFromProto(newFollowerState.PrecommittedAlh)
+
+	followerSt, ok := d.followerStates[newFollowerState.UUID]
+	if ok {
+		if newFollowerState.PrecommittedTxID < followerSt.precommittedTxID {
+			return fmt.Errorf("%w: the newly informed follower state lags behind the previously informed one", ErrIllegalArguments)
+		}
+
+		if newFollowerState.PrecommittedTxID == followerSt.precommittedTxID {
+			// as of the last informed follower status update, nothing has changed
+			return nil
+		}
+
+		// actual replication progress is informed by the follower
+		followerSt.precommittedTxID = newFollowerState.PrecommittedTxID
+		followerSt.precommittedAlh = newFollowerAlh
+	} else {
+		// follower informs first replication state
+		d.followerStates[newFollowerState.UUID] = &followerState{
+			precommittedTxID: newFollowerState.PrecommittedTxID,
+			precommittedAlh:  newFollowerAlh,
+		}
+	}
+
+	// check up to which tx enough followers ack replication and it's safe to commit
+	mayCommitUpToTxID := uint64(0)
+	if len(d.followerStates) > 0 {
+		mayCommitUpToTxID = math.MaxUint64
+	}
+
+	allowances := 0
+
+	// we may clean up followerStates from those who are lag behind commit
+	for _, st := range d.followerStates {
+		if st.precommittedTxID < mayCommitUpToTxID {
+			mayCommitUpToTxID = st.precommittedTxID
+		}
+		allowances++
+	}
+
+	if allowances >= d.options.syncFollowers {
+		err := d.st.AllowCommitUpto(mayCommitUpToTxID)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *db) ExportTxByID(req *schema.ExportTxRequest) (txbs []byte, mayCommitUpToTxID uint64, mayCommitUpToAlh [sha256.Size]byte, err error) {
+	d.exportTxMutex.Lock()
+	defer d.exportTxMutex.Unlock()
```

Comment:
This mutex will remove ability run multiple exports in parallel.

Reply:
at this stage is more about safety, probably the locking may be moved to the state holder